### PR TITLE
Allow `Abort` terminators in all const-contexts

### PR DESCRIPTION
**Pull request by ecstatic-morse**
We never unwind during const-eval, so we basically have these semantics already. Also I just figured out that these only appear along the cleanup path, which doesn't get const-checked. In other words, this doesn't actually change behavior: the `check-pass` test I added compiles just fine on nightly.

r? @RalfJung 
cc @rust-lang/wg-const-eval

### CODE REVIEW

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 1]
```diff
@@ -874,13 +874,13 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
...
 
             TerminatorKind::InlineAsm { .. } => self.check_op(ops::InlineAsm),
-            TerminatorKind::Abort => self.check_op(ops::Abort),
 
             TerminatorKind::GeneratorDrop | TerminatorKind::Yield { .. } => {
                 self.check_op(ops::Generator(hir::GeneratorKind::Gen))
             }
 
-            TerminatorKind::Assert { .. }
+            TerminatorKind::Abort
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:882]: Maybe add a comment here saying that `Abort` terminators are only used for `#[unwind(aborts)]` and since CTFE does not unwind, they are thus unreachable?

File: src/test/ui/consts/unwind-abort.rs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// check-pass
+
+#![feature(unwind_attributes, const_panic)]
+
+// `#[unwind(aborts)]` is okay for a `const fn`. We don't unwind in const-eval anyways.
```
Comment (RalfJung) [inline src/test/ui/consts/unwind-abort.rs:5]: Why did you add two tests? Our test suite is already huge.^^

Comment (ecstatic-morse) [reply src/test/ui/consts/unwind-abort.rs:5]: One is check-pass, the other fails to compile.

File: compiler/rustc_mir/src/transform/check_consts/validation.rs (hunk 1 above)
Comment (ecstatic-morse) [reply compiler/rustc_mir/src/transform/check_consts/validation.rs:882]: I could just `span_bug` here actually. AFAICT, there are only `Abort`s in "cleanup" blocks.

File: src/test/ui/consts/unwind-abort.rs (hunk 2 above)
Comment (RalfJung) [reply src/test/ui/consts/unwind-abort.rs:5]: I can see that, I just do not understand their respective purpose. Specifically, "consts/unwind-abort" does not even run any CTFE.

File: compiler/rustc_mir/src/transform/check_consts/validation.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_consts/validation.rs:882]: That sounds like a good idea but does not replace a comment. :)

File: src/test/ui/consts/unwind-abort.rs (hunk 2 above)
Comment (ecstatic-morse) [reply src/test/ui/consts/unwind-abort.rs:5]: :shrug: I'm happy with the test coverage in this PR.

Comment (RalfJung) [reply src/test/ui/consts/unwind-abort.rs:5]: I am trying to understand what `unwind-abort.rs` tests that the other test does not already cover.

My concern is not too little coverage; my concern is unnecessarily growing the number of UI tests.

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 3]
```diff
@@ -879,8 +881,11 @@ impl Visitor<'tcx> for Validator<'mir, 'tcx> {
                 self.check_op(ops::Generator(hir::GeneratorKind::Gen))
             }
 
-            TerminatorKind::Abort
-            | TerminatorKind::Assert { .. }
+            TerminatorKind::Abort => {
+                span_bug!(self.span, "`Abort` terminator outside of cleanup block")
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:885]:
```suggestion
                // Cleanup blocks are skipped for const checking (see `visit_basic_block_data`).
                span_bug!(self.span, "`Abort` terminator outside of cleanup block")
```

File: src/test/ui/consts/unwind-abort.rs (hunk 2 above)
Comment (ecstatic-morse) [reply src/test/ui/consts/unwind-abort.rs:5]: I try to split up passing and failing tests when possible. One is solely a test of const-checking and involves no const-eval, the other is a sanity check to ensure that the CTFE engine doesn't do anything weird when evaluating something with `#[unwind(aborts)]`.

Comment (RalfJung) [reply src/test/ui/consts/unwind-abort.rs:5]: I think this unnecessarily leads to an excessive number of tests in our UI test suite that is already so big that it becomes hard to run the entire thing, but whatever.

Could you add comments to these tests explaining what they are meant to test?

File: src/test/ui/consts/const-eval/unwind-abort.rs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#![feature(unwind_attributes, const_panic)]
+
+#[unwind(aborts)]
+const fn foo() {
+    panic!() //~ evaluation of constant value failed
+}
+
+const _: () = foo(); //~ any use of this value will cause an error
```
Comment (ecstatic-morse) [inline src/test/ui/consts/const-eval/unwind-abort.rs:8]:
```suggestion
const _: () = foo(); //~ any use of this value will cause an error
// Ensure that the CTFE engine handles calls to `#[unwind(aborts)]` gracefully
```

File: src/test/ui/consts/const-eval/unwind-abort.rs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#![feature(unwind_attributes, const_panic)]
+
+#[unwind(aborts)]
+const fn foo() {
+    panic!() //~ evaluation of constant value failed
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/unwind-abort.rs:5]: Btw, shouldn't this say `//~ ERROR`? I don't think I have seen these comments without a message level before.

Comment (ecstatic-morse) [reply src/test/ui/consts/const-eval/unwind-abort.rs:5]: If you'd like to ban this abbreviated form of test annotation, you're free to do so. I like them because they reduce noise.

Comment (RalfJung) [reply src/test/ui/consts/const-eval/unwind-abort.rs:5]: I'm not proposing to ban anything, I am asking if these are even checking anything, given that the docs do not mention them. If you like them, might be a good idea to add them to the docs.

Comment (ecstatic-morse) [reply src/test/ui/consts/const-eval/unwind-abort.rs:5]: > I am asking if these are even checking anything

They are. They're used all over the test suite. I can document their existence I suppose, but I'm not sure why this responsibility falls on me.

Comment (RalfJung) [reply src/test/ui/consts/const-eval/unwind-abort.rs:5]: Fine then. I did a quick grep before asking you and I guess I didn't notice them in the vast, vast majority of annotations that to explicitly give a level. Given that they are also not documented, I think it is reasonable to ask.
